app/common/src/cljc/teet/util/datomic.cljc

```diff
@@ -199,6 +199,24 @@
        (concat retractions [(assoc changes :db/id id)]))
       [])))
 
+#?(:clj
+   (defn modify-entity-retract-nils
+     "Create transaction data that asserts updated entity map
+     and retracts any attribute whose value is nil.
+
+     If entity :db/id is not a number, no retractions are issued."
+     [db {id :db/id :as new-entity}]
+     (let [entity-values (cu/without-nils new-entity)
+           nil-attributes (into []
+                                (comp
+                                 (filter (comp nil? val))
+                                 (map key))
+                                new-entity)]
+       (into [entity-values]
+             (when (number? id)
+               (for [[attr val] (d/pull db nil-attributes id)]
+                 [:db/retract id attr val]))))))
```

Comment:
As an aside, this could be used to refactor lots of our backend code to something like
```
(-> (pull-entity-from-db entity-id)
    (do-domain-operation other-request-args)
    (persist-using-modify-entity-retract-nils-and-meta-stuff user)
```
where `do-domain-operation` could then be more easily tested in isolation.